Added app name , app version and client capabilities to Auth and Token requests

kreedula opened the pull request:
- Added app name and app version : [API proprosal  ](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview/pullrequest/542?_a=files&path=%2FClientAppNameVersion%2Foverview.md)
- Refactored acquireToken overloads to remove duplicate code.
- Updated  unit tests to validate these new query params

- Added client capabilities support : [API proposal](https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview/pullrequest/533?_a=files&path=%2FClientCapabilities%2Foverview.md)
- Added unit tests to validate client capabilities support.

Todo: 

- Confirm the final format of client capabilities and test with server Test Slice 
- Add sanity checks based on Authenticator and company portal versions to apply these changes only to updated app versions
- Update submodule once https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/280 merges

addev-ashish commented on adal/src/androidTest/java/com/microsoft/aad/adal/Util.java:
```diff
@@ -162,12 +162,16 @@ static String getErrorResponseBody(final String errorCode) throws JSONException
     static byte[] getPoseMessage(final String refreshToken, final String clientId, final String resource)
             throws UnsupportedEncodingException {
-        return String.format("%s=%s&%s=%s&%s=%s&%s=%s&%s=%s",
+        String string = String.format("%s=%s&%s=%s&%s=%s&%s=%s&%s=%s&%s=%s&%s=%s",
```
can we rename the function to getPostMessage

addev-ashish commented on adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java:
```diff
@@ -630,23 +521,47 @@ public void acquireToken(String resource, String clientId, @Nullable String redi
-            createAcquireTokenRequest(apiEvent).acquireToken(null, false, request, callback);
+            setAppInfoToRequest(request);
+
+            if(!TextUtils.isEmpty(loginHint)) {
```
can we use StringExtensions.isNullOrBlank

addev-ashish commented on adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java:
```diff
@@ -630,23 +521,47 @@ public void acquireToken(String resource, String clientId, @Nullable String redi
+                request.setUserIdentifierType(UserIdentifierType.LoginHint);
+            }
+
+            createAcquireTokenRequest(apiEvent).acquireToken(fragment, false, request, callback);
```
second param here is useDialog, some api's like ACQUIRE_TOKEN_7 pass true to it

kreedula replied:
Thanks @addev-ashish for catching this. Updated to add this param as well

addev-ashish commented on adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java:
```diff
@@ -1214,6 +1157,30 @@ public void startActivityForResult(Intent intent, int requestCode) {
         };
     }
 
+    private String mergeClaimsWithClientCapabilities(String claims) throws JSONException {
```
final

addev-ashish replied:
can we change this function to package scope and write some tests to just test the input vs output in various scenarios. that way if we find a bug later we can just add to the unit test.

addev-ashish commented on adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java:
```diff
@@ -1214,6 +1157,30 @@ public void startActivityForResult(Intent intent, int requestCode) {
+            capabilitiesArray.put(capability);
+        }
+
+        JSONObject capabilities = new JSONObject();
```
variables in this function need to be final. also a big comment on the top on what is the formatted output will be nice.

addev-ashish commented on adal/src/main/java/com/microsoft/aad/adal/AuthenticationContext.java:
```diff
@@ -1489,6 +1456,14 @@ public static String getVersionName() {
+    }
+
+    public void setClientCapabilites(List<String> clientCapabilites){
+        this.mClientCapabilites = clientCapabilites;
```
this. is not required

kreedula commented:
Thanks @addev-ashish for the review. Addressed all the comments via https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/1343/commits/ad3bd77b0da379d3cd35629e31ccd106786994f6

kreedula commented:
Thank you @addev-ashish
